### PR TITLE
Fix Sending Deletes on Search Expression Subscription Criterias

### LINKED ISSUE #3703: Specific subscriptions cause NPE on the matcher
Combining the "send deletes OK" extension:
```
http://hapifhir.io/fhir/StructureDefinition/subscription-send-delete-messages
```
with a search expression criteria (e.g. "Patient?gender=male") on a single subscription causes NullPointerExceptions in the SubscriptionMatchingSubscriber. It attempts to match the resource body against the criteria, but there is no resource body, as the payload is intentionally null'ed out in the BaseResourceModifiedMessage.java#L60 in case of delete operation. The following is an example of such a broken subscription: 

```
{
	"resourceType": "Subscription",
	"id": "RestHook",
	"meta": {
		"versionId": "10",
		"lastUpdated": "2022-05-25T11:39:35.346-04:00",
		"tag": [
			{
				"system": "http://hapifhir.io/fhir/StructureDefinition/subscription-matching-strategy",
				"code": "IN_MEMORY",
				"display": "In-memory"
			}
		]
	},
	"status": "active",
	"criteria": "Patient?gender=male",
	"channel": {
		"extension": [
			{
				"url": "http://hapifhir.io/fhir/StructureDefinition/subscription-send-delete-messages",
				"valueBoolean": true
			}
		],
		"type": "rest-hook",
		"endpoint": "https://52d0-216-211-94-24.ngrok.io",
		"payload": "application/json"
	}
}
```

### PR DESCRIPTION
**Pull request by tadgh**
* Modify BaseResourceModifiedMessage to no longer strip the payload body
* Parameterize a test to prove functionality
* Add changelog
* Minor fix for a log message. 
 
 Closes #3703

### CODE REVIEW

**Comment by StevenXLi** [comment]
You might need to check why `testDelete()` was checking if the payload is null in `ResourceModifiedTest`
Some parts of the code might be expecting the payload to be null

**Comment by tadgh** [comment]
> You might need to check why `testDelete()` was checking if the payload is null in `ResourceModifiedTest` Some parts of the code might be expecting the payload to be null

Good catch. looked into it, there's no real reason. It was a space saving measure. Test modified 👍🏻

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceModifiedMessage.java [hunk 1]
```diff
@@ -57,11 +57,7 @@ public BaseResourceModifiedMessage() {
 	public BaseResourceModifiedMessage(FhirContext theFhirContext, IBaseResource theResource, OperationTypeEnum theOperationType) {
 		this();
 		setOperationType(theOperationType);
-		if (theOperationType != OperationTypeEnum.DELETE) {
-			setNewPayload(theFhirContext, theResource);
-		} else {
-			setPayloadIdFromPayload(theFhirContext, theResource);
-		}
+		setNewPayload(theFhirContext, theResource);
```
Comment (fil512) [inline hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceModifiedMessage.java:60]: I preferred the old code here.  But I guess I need to see the ticket to understand the issue here.

**Review by fil512 — APPROVED** [review]

File: hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceModifiedMessage.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceModifiedMessage.java:60]: Ooooh I get it.  Ya this makes sense now.  I guess part of the reason it's weird is that delete subscriptions aren't really even a thing in the spec...

Comment (tadgh) [reply hapi-fhir-server/src/main/java/ca/uhn/fhir/rest/server/messaging/BaseResourceModifiedMessage.java:60]: Yeah i think we did this as a premature optimization for message size, expecting to never need it.

**Pull request merged**